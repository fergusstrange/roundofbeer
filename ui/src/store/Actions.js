const actions = {
  updateError: error => async state => ({
    ...state,
    error,
  }),
  clearError: () => async state => ({
    ...state,
    error: undefined,
  }),
  updateRoundToken: data => async state => ({
    ...state,
    roundToken: data.token,
    round: undefined,
    participatingRounds: Array.isArray(state.participatingRounds)
      ? [...state.participatingRounds.filter(pr => pr.roundUrl !== data.roundUrl),
        {
          roundUrl: data.roundUrl,
          roundToken: data.token,
          participants: data.participants,
        }]
      : [{
        roundUrl: data.roundUrl,
        roundToken: data.token,
        participants: data.participants,
      }],
  }),
  updateRound: round => async state => ({
    ...state,
    round,
  }),
};

export default actions;
