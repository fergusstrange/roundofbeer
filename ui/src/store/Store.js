import { get, set } from 'local-storage';
import { createDakpan } from 'dakpan';
import ApiClient from '../client/Client';

const localStorageKey = 'roundOfBeerLocalStorage';
const client = new ApiClient();

function localStorageStateOrDefaults() {
  return get(localStorageKey)
    || {
      token: undefined,
      round: undefined,
      participant: '',
      participants: [],
    };
}

function updateLocalStore(state) {
  set(localStorageKey, state);
}

const [ContextProvider, useContext] = createDakpan(localStorageStateOrDefaults())({
  clearError: () => async state => ({
    ...state,
    error: undefined,
  }),
  addParticipant: () => ({ participant, ...state }) => ({
    ...state,
    participants: [
      ...state.participants,
      participant,
    ],
    participant: '',
  }),
  updateParticipant: participant => state => ({
    ...state,
    participant,
  }),
  submitParticipants: () => async (state) => {
    const { error, data } = await client.createRound(state.participants);
    return !error
      ? ({
        ...state,
        round: data.round,
        roundToken: data.token,
        participant: '',
        participants: [],
      })
      : ({
        ...state,
        error,
      });
  },
  fetchRound: () => async (state) => {
    const { error, data } = await client.fetchRound(state.roundToken);
    return error
      ? ({
        ...state,
        error,
      })
      : ({
        ...state,
        round: data.round,
        roundToken: data.token,
      });
  },
});

export { ContextProvider, useContext, updateLocalStore };
