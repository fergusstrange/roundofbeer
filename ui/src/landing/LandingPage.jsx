import React from 'react';
import PropTypes from 'prop-types';
import { roundContext } from '../store/Store';

export default function LandingPage({ history }) {
  const [state] = roundContext();

  if (state.round && state.roundToken) {
    history.push(`/${state.round.url}`);
  } else {
    history.push('/new-round');
  }

  return (
    <div>
      You need to enable JavaScript to run this app.
    </div>
  );
}

LandingPage.propTypes = {
  history: PropTypes.shape().isRequired,
};
