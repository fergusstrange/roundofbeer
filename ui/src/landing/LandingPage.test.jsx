import React from 'react';
import renderer from 'react-test-renderer';
import LandingPage from './LandingPage';
import { ContextProvider } from '../store/Store';

describe('Landing Page Tests', () => {
  it('Renders existing round join page', () => expect(renderer.create(
    <ContextProvider>
      <LandingPage
        history={{
          push: () => {},
        }}
      />
    </ContextProvider>,
  ).toJSON())
    .toMatchSnapshot());
});
