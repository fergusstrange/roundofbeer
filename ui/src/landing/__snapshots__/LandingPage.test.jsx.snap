// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing Page Tests Renders existing round join page 1`] = `
<div>
  You need to enable JavaScript to run this app.
</div>
`;
